### DISCUSSION TITLE
How to start MongoDB as a service on Windows runners

### DISCUSSION THREAD

**Discussion by MattiSG** [community/community#30083] [Actions] [answered]
I run a cross-platform test suite in CI relying on a MongoDB server being available.

I understand that MongoDB comes preinstalled in both Linux and [Windows](https://github.com/actions/runner-images/blob/main/images/win/Windows2022-Readme.md#mongodb) images. However, I could find no documentation at all on how to make use of that server.

Since I need to start the `mongod` server and make it available for further steps in my job, I need that step to be done in the background, otherwise the step is blocking. I understand there is no option for such a thing in GitHub Actions.

My first attempt was to use the `--fork` option of `mongod`. However, it is [not supported on Windows](https://www.mongodb.com/docs/manual/reference/program/mongod/#std-option-mongod.--fork) and using `&`, while [supported in PowerShell](https://stackoverflow.com/questions/185575/powershell-equivalent-of-bash-ampersand-for-forking-running-background-proce), [does not seem](https://github.com/ambanum/OpenTermsArchive/actions/runs/2889002081) to start the server since I cannot connect to it afterwards.

I thus wanted to use GitHub Actions’ `services`, but those are [not supported on Windows](https://github.com/actions/runner/issues/904).

Considering that MongoDB comes preinstalled as a service on the runner image, I tried to use `Run Start-Service MongoDB` (or variants like `Run Start-Service -Name "MongoDB"`), but get the following error:

> [Cannot start service 'MongoDB' on computer '.'.](https://github.com/ambanum/OpenTermsArchive/runs/7918828565?check_suite_focus=true)

How can I start the preinstalled MongoDB service on a Windows GitHub Actions runner? Is there any other way for me to run a Mongo server in the background on Windows in a GitHub Actions workflow?

**Comment by AntoineAugusti** [comment, marked as answer]
You may want to check [this issue](https://github.com/actions/runner-images/issues/5949).

```yml
- run: |
        Set-Service mongodb -StartupType Automatic          
        Start-Service -Name MongoDB
```

from this comment https://github.com/actions/runner-images/issues/5949#issuecomment-1210885327

**Reply by MattiSG** [reply]
Thank you so much for this pointer @AntoineAugusti! This is indeed the recommended syntax by GitHub Actions maintainers.

Unfortunately, this solution is flaky: both [our](https://github.com/actions/runner-images/issues/5949#issuecomment-1223828758) and [another](https://github.com/actions/runner-images/issues/5949#issuecomment-1211389442) team encounter seemingly random failures with it. An alternative syntax [has been suggested](https://github.com/actions/runner-images/issues/5949#issuecomment-1223837948) and I am currently trying it out.

**Reply by MattiSG** [reply]
After 10 runs with 3 different syntaxes (details in https://github.com/actions/runner-images/issues/5949#issuecomment-1225522672), the failure rate seems to be around 10% each time, no matter which is used. I thus recommend the one you mentioned in your reply, as it is the most readable. Thank you again for having pointed out this issue, it solved my problem as best as it can be solved 😊

**Reply by jeroen** [reply]
This worked for a while, but now no longer works :( 

```sh
Start-Service: D:\a\_temp\375b21ec-5f27-44d2-b47e-9461a5ad2f4d.ps1:3
Line |
   3 |  Start-Service -Name mongodb
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Service 'MongoDB Server (MongoDB) (mongodb)' cannot be started due to the following error: Cannot start service
     | 'mongodb' on computer '.'.
Error: Process completed with exit code 1.
```

As a workaround you can try to start it repeatedly until success. This works for me:

```yaml
      - if: runner.os == 'windows'
        continue-on-error: true
        run: |
          Set-Service mongodb -StartupType Automatic
          Do {
            try {
              Write-Host "Trying to start MongoDB..."
              Start-Service MongoDB
            }
            catch {
              "Failure occurred in Start-Service MongoDB."
            }
            Start-Sleep -Seconds 5
            $Service = Get-Service MongoDB
          } Until ($Service.Status -eq "Running")
```
